### PR TITLE
Add rules sets list on readme

### LINKED ISSUE #3252: Show rules sets list on readme
The rules sets count is being high.

Just a little thing, but it would be great to have a short list of all rules sets, before the fixers list.

### PR DESCRIPTION
**Pull request by soullivaneuh**
Closes #3252

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -232,6 +232,53 @@ would be default in next MAJOR release (unified differ, estimating, full-width p
 
     $ PHP_CS_FIXER_FUTURE_MODE=1 php php-cs-fixer.phar fix -v --diff
 
+Choose from the list of available sets:
+
+* **@PSR1**
```
Comment (keradus) [inline README.rst]: we need some descriptions for those ;)

Comment (Soullivaneuh) [reply README.rst]: From where can I get the descriptions? It's a generated list.

Comment (keradus) [reply README.rst]: List of rules is also generated.
That's the point, there is no descriptions yet. I must admit that I don't see much value of list having only names.

Comment (Soullivaneuh) [reply README.rst]: > I must admit that I don't see much value of list having only names.

The value is to have a quick tour of which groups we can use while configuring our `.php-cs`, as I described on the issue.

I may add descriptions of each groups if you want, just give me what descriptions you do want.

Comment (keradus) [reply README.rst]: >  just give me what descriptions you do want.

I would say that it's the biggest part to be done ;)
as I said, we don't have them ready anywhere

Comment (Soullivaneuh) [reply README.rst]: I may add some description to see, but the names are clear enough IMHO.

Comment (keradus) [reply README.rst]: so, let us wait for more opinions here ;)

**Comment by julienfalque** [comment]
Maybe we could also add information about rules included in rulesets with a different configuration than the default one. Currently, the documentation only shows which rulesets each rule belongs to, but we have no idea about the configuration that will be used. See #3295.

**Comment by Soullivaneuh** [comment]
@julienfalque I think I can add it with ease on this PR, is it really worth it? The result might be huge.

**Comment by julienfalque** [comment]
> Ultimate solution (out of scope of this PR) is to split README / documentation , as one file was OK when we had ~30 non-configurable rules, not ~160 with multiple config options.
If one wants to help on that, let us catch up on gitter.

I started a POC of a splitted documentation which is published at https://github.com/julienfalque/PHP-CS-Fixer/tree/doc/doc/rules.

**Comment by keradus** [comment]
there is no place to see agenda of proposed change or give a feedback :(

**Comment by SpacePossum** [comment]
I would like to see this finished as we could reuse the RuleSet description in the DescribeCommand.
For example;
```
$ php php-cs-fixer describe @PSR2
Description of @PSR2 set.
```

What is blocking this? I know there are a few awesome ideas about new doc/site generation, but maybe we can a little step here first?

**Comment by Soullivaneuh** [comment]
> What is blocking this? I know there are a few awesome ideas about new doc/site generation, but maybe we can a little step here first?

I agree, the doc improvement ideas looks nice but out of scope of this PR. Adding a dynamic list of rules sets now waiting better displaying won't hurt so much IMHO.

@SpacePossum I may update the PR to add rules sets description. All I need is the description contents of each set.

**Comment by keradus** [comment]
as @SpacePossum would be preparing description of rulesets, I would be eager to see proposal of code to provide them, and I'm not talking about hardcoded text in readme generator

**Comment by Soullivaneuh** [comment]
@keradus Sure, I'll see what I can when I'll have the data. :+1:

**Comment by keradus** [comment]
actually, you can assume `TODO fill me` as a data, and then we will update the data :D

File: src/Console/Command/HelpCommand.php [hunk 2]
```diff
@@ -420,6 +425,16 @@ private static function getChangeLogFile()
...
     }
 
+    /**
+     * @return string
+     */
+    private static function getRulesSetsHelp()
+    {
+        return implode("\n", array_map(function ($setDefinitionName) {
+            return " * <comment>{$setDefinitionName}</comment>";
+        }, (new RuleSet())->getSetDefinitionNames()));
```
Comment (keradus) [inline src/Console/Command/HelpCommand.php]: rulesets shall be sorted

**Comment by Soullivaneuh** [comment]
@keradus Requested change done.

File: src/RuleSetInterface.php [hunk 3]
```diff
@@ -48,6 +48,15 @@ public function getRules();
...
     public function getSetDefinitionNames();
 
+    /**
+     * Get the related description of the set name.
+     *
+     * @param string $setDefinitionName
+     *
+     * @return string
+     */
+    public function getSetDescription($setDefinitionName);
```
Comment (keradus) [inline src/RuleSetInterface.php]: This is public interface, this is BC breaker.

Also, I don't believe that having it here is the best approach. 

I'm aware that having it fully decoupled is a bigger thing to do (like one ruleset definition = one class), are you willing to try do this? If not, let us still decouple descriptions into autonomic, dedicated, internal class

Comment (julienfalque) [reply src/RuleSetInterface.php]: > one ruleset definition = one clas

This might help implementing custom rulesets. Worth the try IMO :+1:

**Comment by julienfalque** [comment]
What do you think about the following descriptions?

* **DoctrineAnnotation**: rules covering Doctrine annotations with configuration based on examples found in [Doctrine Annotation documentation](http://docs.doctrine-project.org/projects/doctrine-common/en/latest/reference/annotations.html) and [Symfony documentation](http://symfony.com/doc/master/bundles/SensioFrameworkExtraBundle/annotations/routing.html)
* **PHP56Migration**: rules to improve code for PHP 5.6 compatibility
* **PHP56Migration:risky**: extension of the **PHP56Migration** ruleset with risky rules
* **PHP70Migration**: rules to improve code for PHP 7.0 compatibility
* **PHP70Migration:risky**: extension of the **PHP70Migration** ruleset with risky rules
* **PHP71Migration**: rules to improve code for PHP 7.1 compatibility
* **PHP71Migration:risky**: extension of the **PHP71Migration** ruleset with risky rules
* **PHPUnit30Migration:risky**: rules to improve tests code for PHPUnit 3.0 compatibility
* **PHPUnit32Migration:risky**: rules to improve tests code for PHPUnit 3.2 compatibility
* **PHPUnit35Migration:risky**: rules to improve tests code for PHPUnit 3.5 compatibility
* **PHPUnit43Migration:risky**: rules to improve tests code for PHPUnit 4.3 compatibility
* **PHPUnit48Migration:risky**: rules to improve tests code for PHPUnit 4.8 compatibility
* **PHPUnit50Migration:risky**: rules to improve tests code for PHPUnit 5.0 compatibility
* **PHPUnit52Migration:risky**: rules to improve tests code for PHPUnit 5.2 compatibility
* **PHPUnit54Migration:risky**: rules to improve tests code for PHPUnit 5.4 compatibility
* **PHPUnit55Migration:risky**: rules to improve tests code for PHPUnit 5.5 compatibility
* **PHPUnit56Migration:risky**: rules to improve tests code for PHPUnit 5.6 compatibility
* **PHPUnit57Migration:risky**: rules to improve tests code for PHPUnit 5.7 compatibility
* **PHPUnit60Migration:risky**: rules to improve tests code for PHPUnit 6.0 compatibility
* **PSR1**: rules that follow [PSR-1](https://www.php-fig.org/psr/psr-1/) standard
* **PSR2**: rules that follow [PSR-2](https://www.php-fig.org/psr/psr-2/) standard
* **Symfony**: rules that follow the official [Symfony Coding Standards](https://symfony.com/doc/current/contributing/code/standards.html)
* **Symfony:risky**: extension of the **Symfony** ruleset with risky rules

Beside adding documentation of rules and their config for each ruleset (https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3256#issuecomment-349306182), I think it would be nice to mention somewhere that fixers are not immutable and may vary with new releases.

**Comment by SpacePossum** [comment]
>that fixers are not immutable and may vary with new releases.

I think that should be `rule sets are not immutable`

**Comment by julienfalque** [comment]
Indeed, thanks ;)

**Comment by keradus** [comment]
I super like @julienfalque proposal for descirptions ;)
can they be taken into consideration into this PR please, @Soullivaneuh ?

**Comment by keradus** [comment]
ping @Soullivaneuh 
any chances to finish this one? ;)

**Comment by soullivaneuh** [comment]
Sorry for the very long reply.

Honestly, this PR is not my priority anymore and I'm not sure to have any time to finish it soon.

However, the PR has allowed edition to any member of this project, if someone is interested to continue. :+1:

**Comment by kubawerlos** [comment]
Does this PR still make any sense since we have https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3475? I think we can close it.

**Comment by julienfalque** [comment]
I don't think #3475 can replace this PR for now: though it adds the list of rulesets a rule is part of, it doesn't provide a dedicated section or page where one can find all rulesets and their purpose.

**Comment by kubawerlos** [comment]
This PR also for now does not provide any information about rulesets.

**Comment by julienfalque** [comment]
Closing as #5216 was merged.

**Pull request closed without merging**